### PR TITLE
[FLINK-24611] Prevent JM from discarding state on checkpoint abortion

### PR DESCRIPTION
**Pull request by rkhachatryan**
This PR depends on: #17179 to use `SharedStateStrategy` in `CompletedCheckpointStore`.

## What is the purpose of the change

Prevent JM from deleting state when a checkpoint is aborted.
For that, reference counts in `SharedStateStrategy` are replaced with `latestUsingCheckpointID`.
Please see JIRA description for more details: FLINK-24611.

## Verifying this change
- Added `CheckpointCoordinatorTest.testSharedStateNotDiscaredOnAbort` to cover the end goal scenario
- Existing tests: `CheckpointCoordinatorTest`, `DefaultCompletedCheckpointStoreTest`, `IncrementalRemoteKeyedStateHandleTest`;

## Brief change log

- Copied from #17179 (please ignore it in review):
`[FLINK-24086][checkpoint] Rebuilding the SharedStateRegistry only when the restore method is called for the first time.`

- Not very significant changes that touch many files:
` [hotfix][runtime] Extract SharedStateRegistry interface `
` [FLINK-24611] Pass checkpoint ID to SharedStateRegistry `

- Mostly changes `CheckpointCoordinator`:
` [FLINK-24611] Register shared state on checkpoint ACK `

- Mostly changes `SharedStateStrategy` and state handles:
` [FLINK-24611] Discard shared state on checkpoint subsumption `

- Mostly changes `CompletedCheckpointStore`:
` [FLINK-24611] Discard shared state on job termination `

- Mostly changes  RocksDB snapshot strategy - should probably be extracted into a separate ticket and PR:
` [FLINK-24611] Don't re-upload unconfirmed RocksDB SST `
(no tests yet since some aspects are being discussed (placeholder))

- Mostly changes `SharedStateStrategy` and state handles:
`[FLINK-24611] Adjust shared state entry de-duplication`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

### CODE REVIEW

**Comment by rkhachatryan** [comment]
@pnowojski could you please take a look at this PR?

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java:66]: Can you move adding the `NO_OP` implementation to the commit that starts using it?

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java]: nit: dIff would be easier to read and less confusing if you kept the original order (`Result` below `SharedStateEntry`) or re-ordered them in separate commit.

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorRestoringTest.java
Comment (pnowojski) [inline flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorRestoringTest.java]: I don't get this change?  It looks like some github/rebasing/merge conflict issue/artefact?

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java:66]: Sure, moved it to `[FLINK-24611] Register shared state on checkpoint ACK`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java]: I didn't re-order them; I removed `Result` and moved `SharedStateEntry` to `SharedStateRegistryImpl` once it wasn't needed in the interface.

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorRestoringTest.java
Comment (rkhachatryan) [reply flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorRestoringTest.java]: Yes, it probably came from the commit from #17179.
I don't see it after rebasing.

**Review by pnowojski — COMMENTED** [review]
Thanks for change and patience waiting for the review. I think code looks quite solid. I still have the last three commits (starting from RocksDB changes) to review. 

Those RocksDB changes might use someone more familiar with this component to take a look as well.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:1211]: In the last commit message:
> For example, if a task sent state for the first time and then failed without completing checkpoint and restarting JM;

Do you mean JM failure? Or TM/task failure?
> For example, if a task sent state for the first time and then JM failed without completing checkpoint and restarting JM

?

Can you elaborate what is the scenario?

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:1094]: No, it isn't.
For consistency, most of the changes would be bundled together (except for refactorings and RocksDB) because they all change some part of a bigger mechanism (at least I don't see a way to split them).

However, I hope that separating these changes eases the review and will also simplify tracking the changes later (unit tests still pass for every commit though).

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/DeactivatedCheckpointCompletedCheckpointStore.java
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/DeactivatedCheckpointCompletedCheckpointStore.java:69]: Yes, the javadoc is wrong (though it wasn't changed in this PR). 
It doesn't sound like it **must** throw an exception instead of NO_OP.

Anyways, I can fix javadoc and also list `getRegistry` as an allowed method in it (I'm afraid that throwing an exception or returning throwing `SharedStateRegistry` can break something).
WDYT?

Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/DeactivatedCheckpointCompletedCheckpointStore.java:69]: Yes, the javadoc is wrong (though it wasn't changed in this PR). 
It doesn't sound like it **must** throw an exception instead of NO_OP.

Anyways, I can fix javadoc and also list `getRegistry` as an allowed method in it (I'm afraid that throwing an exception or returning throwing `SharedStateRegistry` can break something).
WDYT?

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CompletedCheckpointStore.java
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CompletedCheckpointStore.java:117]: That's a merge artefact, sorry, I'll remove one (`getRegistry`).

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java
Comment (rkhachatryan) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java:107]: Yes, I'll update the javadoc. Thanks.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java
Comment (rkhachatryan) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java:268]: Aborted **files** can be re-used, but not `StateHandleID`s held by `materializedSstFiles` (because JM might not remember these keys on abortion).

> edit: actually I think the logic maybe is correct, but something feels inconsistent here. It's really hard for me to name what is inside materializedSstFiles. Those are confirmed files and also those awaiting confirmation. But aborted files are being removed from it.

Right. This logic is the same as before. The purpose of `notifyCheckpointAborted` is purely cleanup. It's called from state backend, so it's not dead code.

The purpose of `lastUploadedSstFiles` is not to handle abortion:
- if a file was not confirmed for any reason, then `StreamStateHandle` must be sent
- if it was, then a placeholder can be sent

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java
Comment (rkhachatryan) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java:102]: Just explained [above](https://github.com/apache/flink/pull/17774#discussion_r767851024) :) 
And also in `PreviousSnapshot.get`.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java
Comment (rkhachatryan) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java:602]: 1. Renaming to `getUploaded`
2. The method has comments inside (it's raison d'etre), I'd like to keep these comments as close to the code as possible. Adding the same as javadoc would add more duplication. So I'd keep only internal ones as it's is private.

WDYT?

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java
Comment (rkhachatryan) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java:504]: I'd like to keep that method separate; there are already several questions about its logic which is not straightforward.

The outer logic (`createUploadFilePaths`) I think is independent of it: re-upload SST if the previous snapshot doesn't have it in some form.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java
Comment (rkhachatryan) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java:485]: Good point!

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:1211]: I meant TM failure (will update the message on rebase).
 
I desribed the scenario in more detail in the ticket: https://issues.apache.org/jira/browse/FLINK-24611?focusedCommentId=17445399&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17445399

**Comment by rkhachatryan** [comment]
Thanks for the review @pnowojski.
I've updated the PR, PTAL at the changes and the responses above.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/DeactivatedCheckpointCompletedCheckpointStore.java
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/DeactivatedCheckpointCompletedCheckpointStore.java]: For the sake of consistency and safety I think it would be better to have it throwing an exception. If this class shouldn't be used (and that's the intention), then better make sure it won't be used at least on the runtime level. Also it seems it should be working fine with throwing version (based on the usages). Or at least not worse than it used to.

For the record, I don't like the throwing version. Ideally with checkpointing disabled, this class should just not be instantiated (for example even having field `SchedulerBase#completedCheckpointStore` as `Optional` or `@Nulalble` would be better), but that's out of scope of this ticket.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java
Comment (pnowojski) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java:271]: >  The purpose of notifyCheckpointAborted is purely cleanup. It's called from state backend, so it's not dead code.

In this context by dead code I meant "not having any meaningful effect". The code would be working just as well without this `notifyCheckpointAborted()`, right? The `materializedSstFiles` would still be cleaned up in `notifyCheckpointComplete()`?

About my proposal of unifying `lastUploadedSstFiles` and `materializedSstFiles` I missed that the former one has also `StreamStateHandle`s. So to unify them, we would need to change `materializedSstFiles` type to: `SortedMap<Long, Map<StateHandleID, StreamStateHandle>>`. (so keeping them separate makes sense).

However I'm still confused by those variables and the inconsistency between them. Once checkpoint is aborted, those sst files are no longer "materialized"? But they are still considered as uploaded? At the very least naming is confusing.

Comment (rkhachatryan) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java:271]: > In this context by dead code I meant "not having any meaningful effect". The code would be working just as well without this notifyCheckpointAborted(), right? The materializedSstFiles would still be cleaned up in notifyCheckpointComplete()?

Got you. Yes, correctness won't be affected by removing this method.

I agree that the naming is confusing. `materializedSstFiles` should probably be called `uploadedStateIDs`. `lastUploadedSstFiles` can then be called `uploadedStateHandles` (I've done this in 144aa68718e831cf07c2daa60b2d33ae149ce592).
WDYT?

I've also tried to combine the two maps into a single `SortedMap<Long, Snapshot> snapshotsByCheckpoint`. But that didn't work because on checkpoint, we might need data from potentially two checkpoints: last uploaded and last confirmed.

Comment (pnowojski) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java:271]: 👍

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:1212]: can you replace the "For example" paragraph with more or less copy paste of this comment? and maybe comment in the code as well? 

Also regarding both commit message and comment:
> without completing checkpoint and restarting JM

maybe rephrase to

> without completing checkpoint and without restarting JM

? As it is, I was confused a bit.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBStateUploader.java
Comment (Myasuka) [inline flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBStateUploader.java:81]:
```suggestion
                        "Uploaded {} ({} bytes) for state handle id: {}",
```

**Review by Myasuka — COMMENTED** [review]
Thanks for creating this PR, I just reviewed the RocksDB related part, please take a look at my comments.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java
Comment (rkhachatryan) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java:104]: > Those artficats uploaded in failed checkpoints are updated but not in this collection.
Do you mean entry removal on checkpoint abortion?

Do you have any other names in mind?

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java
Comment (rkhachatryan) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java:110]: Do you mean that if multiple checkpoint fail contigiously, TM would re-upload already uploaded files?

I think that's not true: once the files are uploaded, handles are added to `lastUploadedSstFiles`.
On abortion, it is **not** cleared.
On checkpoint, JM confirmation is **not** required to re-use these previously uploaded files.
So, `PreviousSnapshot.getUploaded` will return those pre-uploaded files, regardless of confirmation or abortion.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java
Comment (Myasuka) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java:105]: I mean this collection only contains the checkpoint confirmed artifacts. How about `confirmedStateIDs`?

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java
Comment (Myasuka) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java:111]: You are right, the `lastUploadedSstFiles` would only be cleared after calling `createUploadFilePaths`.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java
Comment (rkhachatryan) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java:105]: But aren't they only confirmed upon receiving a notificiation?

Comment (Myasuka) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java:105]: Thanks for the update, current comments LGTM.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java]: You're right; but after rebasing I don't see this re-ordering anymore.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:1212]: I've updated the commit message and commented in code (replaced with scenario from the ticket).
The `restarting JM` part is now gone.

**Comment by rkhachatryan** [comment]
Thanks for the review @pnowojski and @Myasuka !

I've addressed all the comments and rebased updating the commit messages.

Do you have any further feedback?

**Comment by Myasuka** [comment]
The RocksDB related part LGTM.

**Review by pnowojski — APPROVED** [review]
LGTM (assuming conflicts visible above are easy to solve)

**Comment by rkhachatryan** [comment]
Thanks again for reviewing.
I've resolved the conflicts, merging.

**Pull request merged**
